libraries/standard/http/include/http_client.h

```diff
@@ -0,0 +1,339 @@
+ * define HTTPNetworkContext to the type of their network context. This context
+ * is passed into the network interface functions.
+ */
+struct HTTPNetworkContext;
+typedef struct HTTPNetworkContext * HTTPNetworkContext_t;
+
+/**
+ * @brief Transport interface for sending data over the network.
+ * 
+ * If the number of bytes written returned is less than bytesToWrite, then 
+ * #HTTPClient_Send will return HTTP_NETWORK_ERROR. If a negative value is 
+ * returned then this #HTTPClient_Send will also return #HTTP_NETWORK_ERROR.
+ * 
+ * @param[in] context User defined context.
+ * @param[in] pBuffer Buffer to write to the network stack.
+ * @param[in] bytesToWrite Number of bytes to write to the network.
+ * @return The number of bytes written or a negative network error code.
+ */
+typedef int32_t (* HTTPTransportSend_t )( HTTPNetworkContext_t context,
+                                          const void * pBuffer, 
+                                          size_t bytesToWrite );
+
+/**
+ * @brief Transport interface for reading data on the network.
+ * 
+ * This function will read up to bytesToRead amount of data from the network.
+ * If this function returns a value less than zero, then #HTTPClient_Send will
+ * return #HTTP_NETWORK_ERROR. If this function returns less than the 
+ * bytesToRead and greater than zero, then this function will be invoked again
+ * if the data in pBuffer contains a partial HTTP response message.
+ * 
+ * @param[in] context User defined context.
+ * @param[in] pBuffer Buffer to read network data into.
+ * @param[in] bytesToRead Number of bytes requested from the network.
+ * @return The number of bytes read or a negative error code.
+ */
+typedef int32_t (* HTTPTransportRecv_t )( HTTPNetworkContext_t context,
+                                          const void * pBuffer, 
+                                          size_t bytesToRead );
+
+/**
+ * @brief The HTTP Client library transport layer interface.
+ */
+typedef struct HTTPTransportInterface
+{
+    HTTPTransportRecv_t recv;
+    HTTPTransportSend_t send;
+    HTTPNetworkContext_t context;
+} HTTPTransportInterface_t;
+
+/**
+ * @brief The HTTP Client library return status.
+ */
+typedef enum HTTPStatus
+{
+    HTTP_SUCCESS = 0,
+    HTTP_INVALID_PARAMETER,
+    HTTP_NETWORK_ERROR,
+    HTTP_NOT_SUPPORTED,
+    HTTP_PARTIAL_RESPONSE,
+    HTTP_INSUFFICIENT_MEMORY,
+    HTTP_INTERNAL_ERROR,
+    HTTP_SECURITY_ALERT_TOO_MANY_HEADERS,
+    HTTP_SECURITY_ALERT_PARSER_INVALID_CHARACTER,
+    HTTP_SECURITY_ALERT_INVALID_CONTENT_LENGTH,
+    /* TODO: Add return codes as implementation continues. */
+} HTTPStatus_t;
+
+/**
+ * @brief Request headers buffer and how many headers exist in the buffer.
+ */
+typedef struct HTTPRequestHeaders
+{
+    
+    /**
+     * @brief Buffer to hold the raw HTTP request headers.
+     * 
+     * This buffer is not available to modify until #HTTPNetworkInterface_t.recv
```

Comment:
The library is using a shared memory management model similar to the simple taskpool, where I had encountered the same documentation requirement. I think that it would be more powerful in expression to mention that "The buffer memory is owned by the library from initialization (with <API-function>) to sending out the HTTP request (with <HTTPClientSend> function). The buffer should not be modified until by the caller until after the  <ClientSend> function returns.
For space optimization, the user can re-use this buffer for storing the response from the server (in <HTTPClientResponse_t> type).

Reply:
Sure thanks

Reply:
Please note that documentation is still under construction and free to be updated by us after this PR goes in.